Gate correct format usage for fmt.Errorf and klog.*

alexanderConstantinescu opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/ovn-org/ovn-kubernetes/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Trivial changes are exempt from following this template.
If your change is non-trivial, please provide the following information:
-->

**- What this PR does and why is it needed**

This is another spin-off from https://github.com/ovn-org/ovn-kubernetes/pull/978

@girishmg gave me a bunch of valid comments on that PR stating that my log lines and error messages did not adhere to the style guidelines. I saw that we usually never adhere to those, so this PR adds a check that will run on every PR and verify that the following applies:

- `fmt.Errorf("whatever this error is")` starts with a lower case letter
- `klog.*("Whatever this log is")` starts with an upper case letter
- `klog.*("myCamelCase variable name is: %s", myCamelCase)` will remain untouched as it references a variable
- `fmt.Errorf("OVN something something")` will remain untouched - as it's fully capitalized. 
 
I have also used the script and fixed all places in the code that did not adhere to those cases.

PTAL: @girishmg / @trozet 

<!--
A summary of the changes within this pull request and some context
as to why they were made
-->

**- Special notes for reviewers**
<!--
What exactly did you change - you may also defer to information
contained in commit messages. At a bare minimum it's worth highlighting
which areas of the code were changed as it's easier to assign reviewers
-->


**- How to verify it**
<!--
Did you include unit tests? or end-to-end tests?
How can I manually verify that this patch achieves its objective
-->


**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

dave-tucker commented:
This looks great @alexanderConstantinescu. Not sure why it's failing the unit tests.
Possibly worth rebasing as there have been a couple of recent commits that might improve the stability of the tests.

alexanderConstantinescu commented:
> This looks great @alexanderConstantinescu. Not sure why it's failing the unit tests.
> Possibly worth rebasing as there have been a couple of recent commits that might improve the stability of the tests.

It passes unit tests locally on my machine, so I assume it must be a flake.

danwinship commented on go-controller/cmd/ovnkube/ovnkube.go:
```diff
@@ -300,7 +300,7 @@ func watchForChanges(configPath string) error {
 					return
 				}
-				klog.Errorf("fsnotify error %v", err)
+				klog.Errorf("Fsnotify error %v", err)
```
That's wrong (though it's also a terrible log message so being forced to rewrite it would probably be a good thing.)

danwinship commented on go-controller/pkg/ovn/loadbalancer.go:
```diff
@@ -121,7 +121,7 @@ func (ovn *Controller) configureLoadBalancer(lb, sourceIP string, sourcePort int
 	}
 	ovn.setServiceEndpointsToLB(lb, vip, targets)
-	klog.V(5).Infof("lb entry set for %s, %s, %v", lb, lbTarget,
+	klog.V(5).Infof("Lb entry set for %s, %s, %v", lb, lbTarget,
```
LB

danwinship commented on go-controller/cmd/ovnkube/ovnkube.go:
```diff
@@ -228,7 +228,7 @@ func runOvnKube(ctx *cli.Context) error {
 	if master != "" {
 		if runtime.GOOS == "windows" {
-			return fmt.Errorf("Windows is not supported as a master")
+			return fmt.Errorf("windows is not supported as a master")
```
That's wrong; "Windows" is a proper noun.

danwinship commented on go-controller/hybrid-overlay/pkg/controller/node_windows.go:
```diff
@@ -42,15 +42,15 @@ func newNodeController(kube kube.Interface,
 	supportedFeatures := hcn.GetSupportedFeatures()
 	if !supportedFeatures.HostRoute {
-		return nil, fmt.Errorf("This version of windows does not support HostRoute " +
+		return nil, fmt.Errorf("this version of windows does not support HostRoute " +
```
So this is a larger violation of the "errors should be phrase-like, not sentence-like" rule... but changing the first character to be lowercase isn't the right fix. (Likewise the next one.)

alexanderConstantinescu replied:
So, I don't know what to do about that except logging it as `Klog.Error` and add a simpler message for the `fmt.Errorf` (which trickles up to: https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/hybrid-overlay/cmd/hybrid-overlay-node/hybrid-overlay-node.go#L38). Suggestion?

danwinship commented on go-controller/hack/verify-gofmt.sh:
```diff
@@ -6,19 +6,36 @@ set -o pipefail
+rectify_logs=$(echo $find_files \| xargs sed -i -e \''s/\(klog\..*"\)\([a-z]\+\s\)/\1\u\2/g'\')
+
+malformatted_errors=$(eval $find_files | xargs sed -n -e 's/\(fmt\.Errorf.*"\)\([A-Z][a-z]\+\s\)/\1\l\2/p')
+malformatted_logs=$(eval $find_files | xargs sed -n -e 's/\(klog\..*"\)\([a-z]\+\s\)/\1\u\2/p')
```
(Your second commit is fixing the rules you created in the first commit.)

alexanderConstantinescu replied:
Yes, do you want me to squash them and have only one commit which does all of this? I can do that once the PR is reviewed

danwinship replied:
It's fine to split out the klog fixes and the fmt.Errorf fixes into their own commits, but you shouldn't have fixes to `verify-gofmt.sh` mixed in with the klog fixes; that should all be in the first commit

alexanderConstantinescu replied:
Crap, you're right. This change must have been slipped me by during when re-working my commits/code.

girishmg commented:
@alexanderConstantinescu the commit's summary line should also be <= 72 characters. That is another style we have followed and is also 'kind of ' enforced by GitHub, as you can see your first commit's summary message has been truncated. (https://github.com/ovn-org/ovn-kubernetes/pull/1445/commits/5bdda129df7026262210d6ccdcefc0c61f7fd0bb)

alexanderConstantinescu commented:
> Hm... I was initially worried that actually _enforcing_ this would be bad, but it seems like the corrections here are almost entirely correct, so...
> 
> > `klog.*("myCamelCase variable name is: %s", myCamelCase)` will remain untouched as it references a variable
> 
> afaict you did not actually implement that?

Yes, I did, see: https://github.com/alexanderConstantinescu/ovn-kubernetes/blob/format_check/go-controller/pkg/util/node_annotations.go#L227

> @alexanderConstantinescu the commit's summary line should also be <= 72 characters. That is another style we have followed and is also 'kind of ' enforced by GitHub, as you can see your first commit's summary message has been truncated. (5bdda12)

The best way to enforce that is by adding a git hook to the repo, but each developer has to enforce that by him/herself and add it to `.git/hooks/pre-commit`

girishmg commented on go-controller/pkg/ovn/service.go:
```diff
@@ -112,7 +112,7 @@ func (ovn *Controller) syncServices(services []interface{}) {
 	gateways, stderr, err := ovn.getOvnGateways()
 	if err != nil {
-		klog.Errorf("failed to get ovn gateways. Not syncing nodeport"+
+		klog.Errorf("Failed to get ovn gateways. Not syncing nodeport"+
```
Not your issue, but since you are changing this line we need a space at the end -- s/nodeport/nodeport /g

girishmg commented:
> EDIT: besides I am not sure we should add a restriction on a good explanation of what someone's work does, just because GitHub does. We will be interested in reading old git logs to understand why someone did what they did, whereas this restriction is only applied to the title of the PR. Our use case is more important and should thus allow any commit message needed to properly describe the work

@alexanderConstantinescu you could always explain what the commit does in detail in the body but the summary is supposed to be very concise. Please google for` 50/72 git commit rules`. Lot of open source projects enforce it.

alexanderConstantinescu commented:
> @alexanderConstantinescu you could always explain what the commit does in detail in the body but the summary is supposed to be very concise. Please google for` 50/72 git commit rules`. Lot of open source projects enforce it. I understand 50 is way too less, so at least we have been adhering to 72 columns. We should check this through a GitHub action, but for now I have been manually doing it. For the most part, everybody is adhering to this rule of concise git summary line and detailed git body lines.

@girishmg : so I've added a tracked git hook that checks if all commit message lengths are below 50 chars whenever the committer performs a push. I researched how to do this with github's actions, but I didn't manage to find something conclusive and easy. So I've added a make action called: `git-init` which developer's working on this repository can run to activate the git hooks we define in `.githooks`. This will allow us to gate future repository validations in the future....but unfortunately we can't enforce that a developer executes that make action. 

Let me know if you/anyone else has a better idea

alexanderConstantinescu commented:
@girishmg : I removed the commit concerning the title check. PTAL

girishmg commented:
@alexanderConstantinescu Thank you. I will wait for the tests to finish and merge this PR.
